File: apps/elixir/lib/getting_started_ml/chapter12.ex
defmodule GettingStartedMl.Chapter12 do
  @moduledoc """
  第 12 章: 正則化とモデル選択。

  リッジ回帰を Nx で自作し、検証データで正則化の強さを選び、
  `Scholar.Linear.RidgeRegression` と突き合わせる。
  **Scholar にラッソ回帰は無い**ので、座標降下法で自作したものがそのまま最終実装になる。

  行列は「数のリストのリスト」。正則化したモデルは
  `%{coefficients: [列の順に並んだ係数], intercept: 切片}` のマップで表す。
  Elixir のマップはキーの順を保たないので、列の順は必ずリストで持ち回る。
  """

  alias GettingStartedMl.{Chapter02, Chapter07, Dataset}
  alias Scholar.Linear.RidgeRegression

  @feature_columns [:RM, :PTRATIO, :LSTAT]
  @target :PRICE
  @outlier_threshold 3.0

  @test_size 0.3
  @validation_size 0.3
  @seed 0
  @alphas [0.0, 0.1, 1.0, 10.0, 100.0]
  # ラッソ回帰の正則化の強さ。自作の目的関数は件数で割らないので、件数で割る実装
  # （Clojure 版・Java 版が使う Tribuo の ElasticNetCDTrainer）の alpha=0.5 は
  # 訓練データ 47 件では 0.5 * 47 = 23.5 にあたる。
  @lasso_alpha 23.5

  # 座標降下法の繰り返しの上限と、係数の動きの打ち切り。
  @max_iterations 10_000
  @tolerance 1.0e-10

  @doc "特徴量の列。"
  def feature_columns, do: @feature_columns

  @doc "正解の列。"
  def target, do: @target

  @doc "z スコアの絶対値がこの値を超える値を持つ行を外れ値とする。"
  def outlier_threshold, do: @outlier_threshold

  @doc "試す正則化の強さ。"
  def alphas, do: @alphas

  ## リッジ回帰

  @doc """
  特徴量と正解から平均を引いてから `(Xᵀ X + alpha I) w = Xᵀ t` を解いて係数を求め、
  切片は平均値から求める。

  平均を引くのは、切片に罰則をかけないため。`alpha` が 0 なら最小二乗法と同じ解になる。
  """
  def ridge_fit(x, t, alpha) do
    if length(x) != length(t) do
      raise ArgumentError, "特徴量と正解の件数が違います: #{length(x)} と #{length(t)}"
    end

    x_means = column_means(x)
    t_mean = Enum.sum(t) / length(t)
    centered = Nx.tensor(center(x, x_means), type: :f64)
    transposed = Nx.transpose(centered)
    residuals = Nx.new_axis(Nx.tensor(Enum.map(t, &(&1 - t_mean)), type: :f64), 1)

    penalized =
      Nx.add(
        Nx.dot(transposed, centered),
        Nx.multiply(Nx.eye(length(x_means), type: :f64), alpha)
      )

    coefficients =
      penalized
      |> Nx.LinAlg.solve(Nx.dot(transposed, residuals))
      |> Nx.squeeze(axes: [1])
      |> Nx.to_flat_list()

    model(coefficients, t_mean - dot(x_means, coefficients))
  end

  @doc "係数と切片からモデルを作る。"
  def model(coefficients, intercept), do: %{coefficients: coefficients, intercept: intercept}

  @doc "行ごとの予測値。行列の列数は係数の数と同じでなければならない。"
  def predict(%{coefficients: coefficients, intercept: intercept}, x) do
    Enum.map(x, fn row ->
      if length(row) != length(coefficients) do
        raise ArgumentError,
              "特徴量の列数 #{length(row)} と係数の数 #{length(coefficients)} が違います"
      end

      intercept + dot(row, coefficients)
    end)
  end

  @doc "係数の絶対値の合計。正則化が強いほど小さくなる。"
  def coefficient_abs_sum(%{coefficients: coefficients}) do
    coefficients |> Enum.map(&abs/1) |> Enum.sum()
  end

  ## ラッソ回帰（Scholar に無いので自作が最終実装）

  @doc "軟しきい値作用素。`|value|` が `threshold` 以下なら 0 にし、そうでなければ 0 のほうへ縮める。"
  def soft_threshold(value, threshold) when value > threshold, do: value - threshold
  def soft_threshold(value, threshold) when value < -threshold, do: value + threshold
  def soft_threshold(_value, _threshold), do: 0.0

  @doc """
  平均を引いてから座標降下法で `½‖t - Xw‖² + alpha ‖w‖₁` を最小にする。

  L1 の罰則は原点で折れているので、リッジ回帰のように行列を解いて終わりにはできない。
  係数を 1 つずつ順に動かし、軟しきい値作用素で 0 に寄せる。
  """
  def lasso_fit(x, t, alpha) do
    if length(x) != length(t) do
      raise ArgumentError, "特徴量と正解の件数が違います: #{length(x)} と #{length(t)}"
    end

    x_means = column_means(x)
    t_mean = Enum.sum(t) / length(t)
    columns = transpose(center(x, x_means))
    norms = Enum.map(columns, fn column -> dot(column, column) end)
    weights = Tuple.duplicate(0.0, length(columns))

    {fitted, _residuals} =
      descend(
        Enum.zip([columns, norms, 0..(length(columns) - 1)]),
        alpha,
        weights,
        Enum.map(t, &(&1 - t_mean)),
        @max_iterations
      )

    coefficients = Tuple.to_list(fitted)
    model(coefficients, t_mean - dot(x_means, coefficients))
  end

  ## 実験の記録とモデル選択

  @doc "`alpha` ごとにリッジ回帰を訓練データで学習し、訓練データと検証データの決定係数を記録する。"
  def run_ridge_experiments(data, alphas) do
    Enum.map(alphas, fn alpha ->
      fitted = ridge_fit(data.x_train, data.t_train, alpha)

      %{
        alpha: alpha,
        train_score: Chapter07.r2_score(data.t_train, predict(fitted, data.x_train)),
        validation_score: Chapter07.r2_score(data.t_valid, predict(fitted, data.x_valid)),
        coefficient_abs_sum: coefficient_abs_sum(fitted)
      }
    end)
  end

  @doc """
  検証データの決定係数が最も高い実験。同じ値なら先の実験を選ぶ。

  `Enum.max_by/2` は同値なら先のものを返すので、Clojure の `max-key`（同値なら後ろ）と
  逆の心配をしなくてよい。
  """
  def best_experiment([]), do: raise(ArgumentError, "実験結果が 1 件もありません")
  def best_experiment(experiments), do: Enum.max_by(experiments, & &1.validation_score)

  @doc "係数がちょうど 0 になった特徴量の名前を、列の順に返す。"
  def zero_coefficient_names(coefficients, names) do
    if length(coefficients) != length(names) do
      raise ArgumentError, "係数と特徴量名の数が違います"
    end

    coefficients
    |> Enum.zip(names)
    |> Enum.filter(fn {value, _name} -> value == 0.0 end)
    |> Enum.map(&elem(&1, 1))
  end

  ## 標準化して多項式特徴量にする

  @doc "訓練データの列ごとの平均と、件数 n で割る標準偏差を求める。"
  def scaler_fit(x, columns) do
    stats =
      Enum.map(columns, fn column ->
        values = Enum.map(x, &Map.fetch!(&1, column))
        mean = Enum.sum(values) / length(values)
        {mean, :math.sqrt(sum_of_squares(Enum.map(values, &(&1 - mean))) / length(values))}
      end)

    %{
      columns: columns,
      means: Enum.map(stats, &elem(&1, 0)),
      stds: Enum.map(stats, &elem(&1, 1))
    }
  end

  @doc ~S"""
  変換後の列名。元の列、2 乗の列（`"RM^2"`）、積の列（`"RM LSTAT"`）の順。
  """
  def feature_names(%{columns: columns}) do
    names = Enum.map(columns, &to_string/1)

    names ++
      Enum.map(pairs(length(columns)), fn
        {i, i} -> "#{Enum.at(names, i)}^2"
        {i, j} -> "#{Enum.at(names, i)} #{Enum.at(names, j)}"
      end)
  end

  @doc "標準化した値と、その 2 次の項を並べた行列にする。"
  def scaler_transform(scaler, x) do
    %{columns: columns, means: means, stds: stds} = scaler
    combinations = pairs(length(columns))

    Enum.map(x, fn features ->
      z =
        Enum.zip([columns, means, stds])
        |> Enum.map(fn {column, mean, std} -> (Map.fetch!(features, column) - mean) / std end)

      z ++ Enum.map(combinations, fn {i, j} -> Enum.at(z, i) * Enum.at(z, j) end)
    end)
  end

  ## 外れ値を除く

  @doc """
  列ごとの z スコア（標準偏差は件数 n - 1 で割る標本標準偏差）の絶対値が
  `threshold` を超える値を 1 つでも持つ行を除く。
  """
  def remove_outliers(table, columns, threshold) do
    stats =
      Enum.map(columns, fn column ->
        values = Enum.map(table.rows, &required_number(&1, column))
        mean = Enum.sum(values) / length(values)

        {column, mean,
         :math.sqrt(sum_of_squares(Enum.map(values, &(&1 - mean))) / (length(values) - 1))}
      end)

    %{table | rows: Enum.reject(table.rows, &outlier?(&1, stats, threshold))}
  end

  ## Scholar のリッジ回帰

  @doc """
  `Scholar.Linear.RidgeRegression` で学習し、自作と同じ形のモデルにする。

  Scholar のリッジ回帰が最小にするのは `‖y - Xw‖² + alpha ‖w‖²` で、自作と同じ尺度。
  `fit_intercept?` は既定で `true` なので、Scholar も平均を引いてから解き、切片に罰則をかけない。

  `solver` は既定の `:svd`（特異値分解）と `:cholesky`（`Nx.LinAlg.solve` による閉形式）を選べる。
  自作と同じ解き方は `:cholesky` のほうで、そちらなら 1 ビットも違わない。
  """
  def scholar_ridge_fit(x, t, alpha, solver \\ :svd) do
    fitted =
      RidgeRegression.fit(Nx.tensor(x, type: :f64), Nx.tensor(t, type: :f64),
        alpha: alpha,
        solver: solver
      )

    model(
      Nx.to_flat_list(fitted.coefficients),
      Nx.to_number(Nx.squeeze(fitted.intercept))
    )
  end

  ## ボストンの住宅価格

  @doc """
  外れ値を除き、全体を訓練用とテスト用に、訓練用をさらに訓練データと検証データに分ける。

  標準化の平均と標準偏差は訓練データだけから求め、3 つとも同じ値で変換する。
  """
  def prepare_boston(path, test_size, validation_size, seed) do
    table =
      remove_outliers(
        Chapter02.load_table(path),
        [@target | @feature_columns],
        @outlier_threshold
      )

    x = Enum.map(table.rows, &feature_map/1)
    t = Enum.map(table.rows, &required_number(&1, @target))
    outer = Chapter02.split_train_test(x, t, test_size, seed)
    inner = Chapter02.split_train_test(outer.x_train, outer.t_train, validation_size, seed)
    scaler = scaler_fit(inner.x_train, @feature_columns)

    %{
      x_train: scaler_transform(scaler, inner.x_train),
      t_train: inner.t_train,
      x_valid: scaler_transform(scaler, inner.x_test),
      t_valid: inner.t_test,
      x_test: scaler_transform(scaler, outer.x_test),
      t_test: outer.t_test,
      feature_names: feature_names(scaler),
      kept: length(table.rows)
    }
  end

  ## 実データでの実行

  @doc "正則化の強さごとにリッジ回帰を学習し、検証データで選んだモデルとラッソ回帰の結果を表示する。"
  def run do
    path = Path.join(Dataset.dir(), "Boston.csv")
    total = length(Chapter02.load_table(path).rows)
    data = prepare_boston(path, @test_size, @validation_size, @seed)
    experiments = run_ridge_experiments(data, @alphas)
    best = best_experiment(experiments)

    print_summary(data, total)
    print_experiments(experiments)
    IO.puts("検証データで選んだ alpha: #{best.alpha}")
    print_test_scores(data, best.alpha)
    print_lasso(data)
  end

  defp print_summary(data, total) do
    IO.puts("データ件数: #{data.kept}（外れ値 #{total - data.kept} 件を除外）")

    IO.puts(
      "訓練データ: #{length(data.t_train)} 件, 検証データ: #{length(data.t_valid)} 件, " <>
        "テストデータ: #{length(data.t_test)} 件"
    )

    IO.puts("特徴量: #{Enum.join(data.feature_names, ", ")}")
  end

  defp print_experiments(experiments) do
    IO.puts("alpha  訓練 R²  検証 R²  係数の絶対値の合計")

    Enum.each(experiments, fn experiment ->
      IO.puts(
        "#{pad(format(experiment.alpha, 1), 5)}  #{format(experiment.train_score, 4)}  " <>
          "#{format(experiment.validation_score, 4)}  #{format(experiment.coefficient_abs_sum, 3)}"
      )
    end)
  end

  defp print_test_scores(data, alpha) do
    linear = ridge_fit(data.x_train, data.t_train, 0.0)
    ridge = ridge_fit(data.x_train, data.t_train, alpha)
    library = scholar_ridge_fit(data.x_train, data.t_train, alpha)

    IO.puts(
      "テストデータの決定係数: 線形回帰 #{format(test_score(linear, data), 4)}, " <>
        "リッジ回帰 #{format(test_score(ridge, data), 4)}"
    )

    IO.puts("Scholar のリッジ回帰との係数の最大の差: #{exponent(max_difference(ridge, library))}")
  end

  defp print_lasso(data) do
    lasso = lasso_fit(data.x_train, data.t_train, @lasso_alpha)
    zeros = zero_coefficient_names(lasso.coefficients, data.feature_names)

    IO.puts("ラッソ回帰（alpha=#{@lasso_alpha}）で係数が 0 になった特徴量: #{Enum.join(zeros, ", ")}")
  end

  defp test_score(fitted, data), do: Chapter07.r2_score(data.t_test, predict(fitted, data.x_test))

  defp max_difference(left, right) do
    left.coefficients
    |> Enum.zip_with(right.coefficients, fn a, b -> abs(a - b) end)
    |> Enum.max()
  end

  defp feature_map(row) do
    Map.new(@feature_columns, fn column -> {column, required_number(row, column)} end)
  end

  defp required_number(row, column) do
    Chapter02.number(row, column) || raise(ArgumentError, "値が空欄です: #{column}")
  end

  defp outlier?(row, stats, threshold) do
    Enum.any?(stats, fn {column, mean, std} ->
      abs((required_number(row, column) - mean) / std) > threshold
    end)
  end

  # 係数がどれも @tolerance より動かなくなるまで、すべての係数を 1 回ずつ動かす。
  defp descend(_columns, _alpha, weights, residuals, 0), do: {weights, residuals}

  defp descend(columns, alpha, weights, residuals, iterations) do
    {weights, residuals, delta} = sweep(columns, alpha, weights, residuals)

    if delta < @tolerance do
      {weights, residuals}
    else
      descend(columns, alpha, weights, residuals, iterations - 1)
    end
  end

  # すべての係数を 1 回ずつ動かし、いちばん大きく動いた大きさを一緒に返す。
  defp sweep(columns, alpha, weights, residuals) do
    Enum.reduce(columns, {weights, residuals, 0.0}, fn entry, acc ->
      update(entry, alpha, acc)
    end)
  end

  defp update({_column, +0.0, _index}, _alpha, acc), do: acc

  defp update({column, norm, index}, alpha, {weights, residuals, delta}) do
    old = elem(weights, index)

    # いったん列の寄与を残差に戻してから、残差との相関で係数を決め直す
    restored = Enum.zip_with(residuals, column, fn r, value -> r + old * value end)
    weight = soft_threshold(dot(column, restored), alpha) / norm
    removed = Enum.zip_with(restored, column, fn r, value -> r - weight * value end)

    {put_elem(weights, index, weight), removed, max(delta, abs(weight - old))}
  end

  # 2 次の項を作る列の組（i <= j）。
  defp pairs(n), do: for(i <- 0..(n - 1), j <- i..(n - 1), do: {i, j})

  defp column_means([first | _] = x) do
    Enum.map(0..(length(first) - 1), fn j ->
      Enum.sum(Enum.map(x, &Enum.at(&1, j))) / length(x)
    end)
  end

  defp center(x, means), do: Enum.map(x, fn row -> Enum.zip_with(row, means, &(&1 - &2)) end)

  defp transpose(x), do: Enum.zip_with(x, & &1)

  defp dot(a, b), do: Enum.zip_with(a, b, &(&1 * &2)) |> Enum.sum()

  defp sum_of_squares(values), do: Enum.sum(Enum.map(values, &(&1 * &1)))

  defp pad(text, width), do: String.pad_leading(text, width)

  # :io_lib.format はロケールに依らないので、小数点は常に「.」になる。
  defp format(value, digits) do
    ~c"~.#{digits}f" |> :io_lib.format([value]) |> to_string()
  end

  defp exponent(value) do
    ~c"~.2e" |> :io_lib.format([value]) |> to_string()
  end
end

File: apps/clojure/src/getting_started_ml/chapter12.clj
(ns getting-started-ml.chapter12
  "第 12 章: 正則化とモデル選択。リッジ回帰を自作し、検証データで正則化の強さを選び、
   Tribuo の ElasticNetCDTrainer のラッソ回帰・リッジ回帰と突き合わせる。"
  (:require [clojure.string :as str]
            [getting-started-ml.chapter02 :as chapter02]
            [getting-started-ml.chapter07.matrix :as matrix]
            [getting-started-ml.chapter07.metrics :as metrics]
            [getting-started-ml.chapter07.tribuo :as tribuo]
            [getting-started-ml.dataset :as dataset])
  (:import [java.util.logging Level Logger]
           [org.tribuo.math.la SparseVector]
           [org.tribuo.regression.slm ElasticNetCDTrainer SparseLinearModel]))

;; 行列は第 7 章と同じ「ベクタのベクタ」。正則化したモデルは
;; {:coefficients [列の順に並んだ係数] :intercept 切片} のマップで表す。

;; ## 第 7 章の行列に足りない演算

(defn matrix-plus
  "同じ大きさの行列の和。"
  [a b]
  (when-not (and (= (matrix/row-count a) (matrix/row-count b))
                 (= (matrix/column-count a) (matrix/column-count b)))
    (throw (IllegalArgumentException.
            (str (matrix/row-count a) " 行 " (matrix/column-count a) " 列の行列と "
                 (matrix/row-count b) " 行 " (matrix/column-count b) " 列の行列は足せません"))))
  (mapv #(mapv + %1 %2) a b))

(defn matrix-scale
  "すべての成分を scalar 倍した行列。"
  [scalar m]
  (mapv (fn [row] (mapv #(* scalar %) row)) m))

(defn identity-matrix
  "size 行 size 列の単位行列。"
  [size]
  (mapv (fn [i] (mapv #(if (= i %) 1.0 0.0) (range size))) (range size)))

;; ## リッジ回帰

(defn- column-means
  "列ごとの平均。"
  [m]
  (mapv (fn [j] (/ (reduce + (matrix/column m j)) (matrix/row-count m)))
        (range (matrix/column-count m))))

(defn- center
  "列ごとに平均を引いた行列。"
  [m means]
  (mapv (fn [row] (mapv - row means)) m))

(defn ridge-fit
  "特徴量と正解から平均を引いてから (Xᵀ X + alpha I) w = Xᵀ t を解いて係数を求め、切片は平均値から求める。
   平均を引くのは、切片に罰則をかけないため。alpha が 0 なら最小二乗法と同じ解になる。"
  [x t alpha]
  (when-not (= (matrix/row-count x) (count t))
    (throw (IllegalArgumentException.
            (str "特徴量と正解の件数が違います: " (matrix/row-count x) " と " (count t)))))
  (let [x-means (column-means x)
        t-mean (/ (reduce + t) (count t))
        xc (center x x-means)
        tc (matrix/column-vector (mapv #(- % t-mean) t))
        xct (matrix/transpose xc)
        penalized (matrix-plus (matrix/multiply xct xc)
                               (matrix-scale alpha (identity-matrix (count x-means))))
        coefficients (matrix/column (matrix/solve penalized (matrix/multiply xct tc)) 0)]
    {:coefficients coefficients
     :intercept (- t-mean (reduce + (map * x-means coefficients)))}))

(defn predict
  "行ごとの予測値。行列の列数は係数の数と同じでなければならない。"
  [{:keys [coefficients intercept]} x]
  (when-not (= (matrix/column-count x) (count coefficients))
    (throw (IllegalArgumentException.
            (str "特徴量の列数 " (matrix/column-count x) " と係数の数 " (count coefficients) " が違います"))))
  (mapv (fn [row] (reduce + intercept (map * row coefficients))) x))

(defn coefficient-abs-sum
  "係数の絶対値の合計。正則化が強いほど小さくなる。"
  [{:keys [coefficients]}]
  (reduce + (map abs coefficients)))

;; ## 実験の記録とモデル選択

(defn run-ridge-experiments
  "alpha ごとにリッジ回帰を訓練データで学習し、訓練データと検証データの決定係数を記録する。"
  [x-train t-train x-valid t-valid alphas]
  (mapv (fn [alpha]
          (let [model (ridge-fit x-train t-train alpha)]
            {:alpha alpha
             :train-score (metrics/r2-score t-train (predict model x-train))
             :validation-score (metrics/r2-score t-valid (predict model x-valid))
             :coefficient-abs-sum (coefficient-abs-sum model)}))
        alphas))

(defn best-experiment
  "検証データの決定係数が最も高い実験。同じ値なら先の実験を選ぶ。
   max-key は同点なら後ろを返すので、厳密な不等号で畳む。"
  [experiments]
  (when (empty? experiments)
    (throw (IllegalArgumentException. "実験結果が 1 件もありません")))
  (reduce (fn [best experiment]
            (if (> (:validation-score experiment) (:validation-score best)) experiment best))
          experiments))

(defn zero-coefficient-names
  "係数がちょうど 0 になった特徴量の名前を、列の順に返す。"
  [coefficients feature-names]
  (when-not (= (count coefficients) (count feature-names))
    (throw (IllegalArgumentException. "係数と特徴量名の数が違います")))
  (vec (keep (fn [[value name*]] (when (zero? value) name*))
             (map vector coefficients feature-names))))

;; ## 標準化して多項式特徴量にする

(defn scaler-fit
  "訓練データの列ごとの平均と、件数 n で割る標準偏差を求める。"
  [x columns]
  (let [stats (mapv (fn [column]
                      (let [values (mapv #(get % column) x)
                            mean (/ (reduce + values) (count values))]
                        [mean (Math/sqrt (/ (reduce + (map #(let [d (- % mean)] (* d d)) values))
                                            (count values)))]))
                    columns)]
    {:columns (vec columns) :means (mapv first stats) :stds (mapv second stats)}))

(defn- pairs
  "2 次の項を作る列の組（i <= j）。"
  [n]
  (vec (for [i (range n) j (range i n)] [i j])))

(defn feature-names
  "変換後の列名。元の列、2 乗の列（\"RM^2\"）、積の列（\"RM LSTAT\"）の順。"
  [{:keys [columns]}]
  (into (mapv name columns)
        (map (fn [[i j]]
               (if (= i j)
                 (str (name (columns i)) "^2")
                 (str (name (columns i)) " " (name (columns j))))))
        (pairs (count columns))))

(defn scaler-transform
  "標準化した値と、その 2 次の項を並べた行列にする。"
  [{:keys [columns means stds]} x]
  (let [ps (pairs (count columns))]
    (mapv (fn [features]
            (let [z (mapv (fn [column mean std] (/ (- (get features column) mean) std))
                          columns means stds)]
              (into z (map (fn [[i j]] (* (z i) (z j)))) ps)))
          x)))

;; ## 外れ値を除く

(def outlier-threshold
  "z スコアの絶対値がこの値を超える値を持つ行を外れ値とする。"
  3.0)

(defn remove-outliers
  "列ごとの z スコア（標準偏差は件数 n - 1 で割る標本標準偏差）の絶対値が
   threshold を超える値を 1 つでも持つ行を除く。"
  [table columns threshold]
  (let [stats (mapv (fn [column]
                      (let [values (mapv #(chapter02/number % column) (:rows table))
                            mean (/ (reduce + values) (count values))]
                        [column mean
                         (Math/sqrt (/ (reduce + (map #(let [d (- % mean)] (* d d)) values))
                                       (dec (count values))))]))
                    columns)]
    (update table :rows
            (fn [rows]
              (vec (remove (fn [row]
                             (some (fn [[column mean std]]
                                     (> (abs (/ (- (chapter02/number row column) mean) std))
                                        threshold))
                                   stats))
                           rows))))))

;; ## Tribuo の ElasticNetCDTrainer

(def ^:private min-l1-ratio
  "ElasticNetCDTrainer は l1Ratio が 0（純粋なリッジ回帰）を受け付けないので、その代わりに使う値。"
  1e-12)

(def ^:private tolerance 1e-10)
(def ^:private max-iterations 100000)

(defn- to-features
  "行列を、列名を付けた特徴量のマップにする。"
  [x tribuo-columns]
  (mapv (fn [row] (zipmap tribuo-columns row)) x))

(defn tribuo-columns
  "Tribuo に渡すときの列名。Tribuo は列名の順に特徴量を並べ替えるので、桁をそろえた名前にする。"
  [n]
  (mapv #(keyword (format "x%02d" %)) (range n)))

(defn fit-elastic-net
  "ElasticNetCDTrainer で学習し、係数と切片を取り出す。
   Tribuo は特徴量の平均を引いてから学習するので、切片は特徴量と正解の平均値から求める。"
  [x t alpha l1-ratio]
  (let [columns (tribuo-columns (matrix/column-count x))
        trainer (ElasticNetCDTrainer. (double alpha) (double l1-ratio) (double tolerance)
                                      (int max-iterations) false 0)
        ^SparseLinearModel model (tribuo/train trainer (to-features x columns) t columns)
        ^SparseVector weights (first (vals (.getWeights model)))
        coefficients (mapv (fn [column]
                             (.get weights (.getID (.get (.getFeatureIDMap model) (name column)))))
                           columns)
        x-means (column-means x)]
    {:coefficients coefficients
     :intercept (- (/ (reduce + t) (count t)) (reduce + (map * x-means coefficients)))}))

(defn fit-lasso
  "l1Ratio を 1 にしたラッソ回帰。"
  [x t alpha]
  (fit-elastic-net x t alpha 1.0))

(defn fit-ridge
  "自作のリッジ回帰と同じ alpha の尺度で、ElasticNetCDTrainer にリッジ回帰を学習させる。
   ElasticNetCDTrainer は誤差の 2 乗の合計を 2n で割った値に罰則を足すので、alpha を件数 n で割って渡す。"
  [x t alpha]
  (fit-elastic-net x t (/ alpha (count t)) min-l1-ratio))

;; ## ボストンの住宅価格

(def feature-columns
  "特徴量の列。"
  [:RM :PTRATIO :LSTAT])

(def target
  "正解の列。"
  :PRICE)

(defn prepare-boston
  "外れ値を除き、全体を訓練用とテスト用に、訓練用をさらに訓練データと検証データに分ける。
   標準化の平均と標準偏差は訓練データだけから求め、3 つとも同じ値で変換する。"
  [csv-file test-size validation-size seed]
  (let [table (remove-outliers (chapter02/load-table csv-file)
                               (conj feature-columns target) outlier-threshold)
        x (mapv (fn [row] (into {} (map (fn [column] [column (chapter02/number row column)]))
                                feature-columns))
                (:rows table))
        t (mapv #(chapter02/number % target) (:rows table))
        outer (chapter02/split-train-test x t test-size seed)
        inner (chapter02/split-train-test (:x-train outer) (:t-train outer) validation-size seed)
        scaler (scaler-fit (:x-train inner) feature-columns)]
    {:x-train (scaler-transform scaler (:x-train inner)) :t-train (:t-train inner)
     :x-valid (scaler-transform scaler (:x-test inner)) :t-valid (:t-test inner)
     :x-test (scaler-transform scaler (:x-test outer)) :t-test (:t-test outer)
     :feature-names (feature-names scaler)
     :kept (count (:rows table))}))

;; ## 実データでの実行

(def ^:private test-size 0.3)
(def ^:private validation-size 0.3)
(def ^:private seed 0)
(def ^:private alphas [0.0 0.1 1.0 10.0 100.0])
(def ^:private lasso-alpha 0.5)

(defn run
  "正則化の強さごとにリッジ回帰を学習し、検証データで選んだモデルとラッソ回帰の結果を表示する。"
  []
  ;; Tribuo の座標降下法が収束のたびに出す INFO のログを表示しない
  (.setLevel (Logger/getLogger (.getName ElasticNetCDTrainer)) Level/WARNING)
  (let [csv-file (str (dataset/dir) "/Boston.csv")
        total (count (:rows (chapter02/load-table csv-file)))
        {:keys [x-train t-train x-valid t-valid x-test t-test kept] :as data}
        (prepare-boston csv-file test-size validation-size seed)
        experiments (run-ridge-experiments x-train t-train x-valid t-valid alphas)
        best (best-experiment experiments)
        linear (ridge-fit x-train t-train 0.0)
        ridge (ridge-fit x-train t-train (:alpha best))
        lasso (fit-lasso x-train t-train lasso-alpha)]
    (println (str "データ件数: " kept "（外れ値 " (- total kept) " 件を除外）"))
    (println (str "訓練データ: " (count t-train) " 件, 検証データ: " (count t-valid)
                  " 件, テストデータ: " (count t-test) " 件"))
    (println (str "特徴量: " (str/join ", " (:feature-names data))))
    (println "alpha  訓練 R²  検証 R²  係数の絶対値の合計")
    (doseq [{:keys [alpha train-score validation-score coefficient-abs-sum]} experiments]
      (println (format "%5.1f  %.4f  %.4f  %.3f"
                       alpha train-score validation-score coefficient-abs-sum)))
    (println (str "検証データで選んだ alpha: " (:alpha best)))
    (println (format "テストデータの決定係数: 線形回帰 %.4f, リッジ回帰 %.4f"
                     (metrics/r2-score t-test (predict linear x-test))
                     (metrics/r2-score t-test (predict ridge x-test))))
    (println (str "ラッソ回帰（alpha=" lasso-alpha "）で係数が 0 になった特徴量: "
                  (str/join ", " (zero-coefficient-names (:coefficients lasso)
                                                         (:feature-names data)))))))
